### PR TITLE
Weam

### PR DESCRIPTION
**Pull request by mphprogrammer**
DS Redux. To test, log in and upload the weams csv via (CSV Files -> new). The file has 52000+ rows and should take 2.5 minutes to upload and process. You can see the file upload record at (CSV Files -> list) or (CSV Types -> list -> show). Next go to Intermediate CSV Data -> Weam to see the populated weam table. add/create/destroy/update table as desired.

NOTE: The table size (52000+ rows) for a weams CSV exceeds the Heroku limit. We cannot pare down tables, because the csvs are interdependent.

### CODE REVIEW

**Comment by mphprogrammer** [comment]
[WEAMS Extract 20160202.csv.zip](https://github.com/department-of-veterans-affairs/gibct-data-service/files/175540/WEAMS.Extract.20160202.csv.zip)

**Comment by mphprogrammer** [comment]
As far as the travis rspec errors, these were not errors on my dev environment, but only on travis. After putzing around (putzing wasn't flagged as misspelled, interesting). The application defaults to EST, and all PG sql datetime attributes are updated accordingly. There is a class scope method called latest_upload_date, which on my machine returned EST. However on travis, latest_upload_date returned GMT causing the test to fail on travis (and still passed on my machine). I used in_time_zone to convert GMT to EST. However on travis, in_time_zone and a DateTime attribute on a record have different precisions, causing the test to fail. This was not a problem on my machine as both returned the same precision (6 places - 1 microsecond). By comparing them using strftime with 6 place precision the tests now pass on travis.

File: app/controllers/csv_files_controller.rb [hunk 1]
```diff
@@ -0,0 +1,109 @@
+class CsvFilesController < ApplicationController
+  include Alertable
+
+  # Kluge for brakeman
+  STI = {
```
Comment (plusjeff) [inline app/controllers/csv_files_controller.rb]: What does this do?

File: app/helpers/csv_files_helper.rb [hunk 2]
```diff
@@ -0,0 +1,2 @@
+module CsvFilesHelper
```
Comment (plusjeff) [inline app/helpers/csv_files_helper.rb]: Is this file needed? (nothing in it)

File: app/helpers/csv_types_helper.rb [hunk 3]
```diff
@@ -0,0 +1,2 @@
+module CsvTypesHelper
```
Comment (plusjeff) [inline app/helpers/csv_types_helper.rb]: Is this file needed (nothing in it)?

File: app/helpers/weams_helper.rb [hunk 4]
```diff
@@ -0,0 +1,2 @@
+module WeamsHelper
```
Comment (plusjeff) [inline app/helpers/weams_helper.rb]: What does this file do?

File: app/models/csv_files/csv_file.rb [hunk 5]
```diff
@@ -0,0 +1,147 @@
+class CsvFile < ActiveRecord::Base
+  attr_accessor :upload
+
+  STI = %W(WeamsCsvFile)
```
Comment (plusjeff) [inline app/models/csv_files/csv_file.rb]: What is STI? 

This also looks like a list that appeared elsewhere in another module. If this is the case, can it be centralized?

File: app/views/csv_files/_csv_file.html.erb [hunk 6]
```diff
@@ -0,0 +1,18 @@
...
+	<td><%= csv_file.id %></td>
+	<td><%= csv_file.name %></td>
+	<td><%= csv_file.humanize_type %></td>
+	<td><%= csv_file.humanize_date %></td>
+
+  <td>
+  	<%= link_to 'Show', csv_file %>
+  </td>
+  <td>
+  	<%= link_to 'Destroy', csv_file, method: :delete, data: { confirm: 'Are you sure?' } %>
```
Comment (plusjeff) [inline app/views/csv_files/_csv_file.html.erb]: Design nitpick: I remember a conversation with @rickleegit or someone else on not calling this function `Destroy`. Was another word settled upon?

File: app/views/csv_files/_form.html.erb [hunk 7]
```diff
@@ -0,0 +1,29 @@
+<%= form_for(@csv_file, multipart: true) do |f| %>
```
Comment (plusjeff) [inline app/views/csv_files/_form.html.erb]: (Note to myself: Leaving off here)

**Comment by jeffmaher** [comment]
This is a **giant** pull request. Got about 1/3 of the way (left myself a comment on where I left off). Will return to this first thing tomorrow morning.

If you build any future PRs off this branch, just leave a note that this PR needs to be finished first.

File: app/controllers/csv_files_controller.rb (hunk 1 above)
Comment (mphprogrammer) [inline app/controllers/csv_files_controller.rb]: Alertable just produces an unsigned list of action errors, and is used by every controller.

File: app/helpers/csv_files_helper.rb (hunk 2 above)
Comment (mphprogrammer) [inline app/helpers/csv_files_helper.rb]: Nope, just an artifact of rails g resource. Will remove ...

File: app/helpers/csv_types_helper.rb (hunk 3 above)
Comment (mphprogrammer) [inline app/helpers/csv_types_helper.rb]: Nope, just an artifact of rails g resource. Will remove ...

File: app/helpers/weams_helper.rb (hunk 4 above)
Comment (mphprogrammer) [inline app/helpers/weams_helper.rb]: Nothing, just an artifact of rails g resource before I changed the app around. Will remove ...

File: app/models/csv_files/csv_file.rb (hunk 5 above)
Comment (mphprogrammer) [inline app/models/csv_files/csv_file.rb]: There are two STI, one in the csv_file_controller that was a leftover from the older version that is no longer needed (was a workaround for satisfy specs and  brakeman warnings with Single Table Inheritance), I will remove that. The other is required for validation and selecting and associated table containing the processed data of a csv. There will only be one.

File: app/views/csv_files/_csv_file.html.erb (hunk 6 above)
Comment (mphprogrammer) [inline app/views/csv_files/_csv_file.html.erb]: I think I recall that, only I am not sure what we settled on. I think it was in the context of the destroy would not be visible to non "admin" users, though there are no longer roles in the immediate future. I really am not sure what we settled on.

**Comment by mphprogrammer** [comment]
Regarding the size, I am sorry for that. Not all of the changes were code changes - About 25% of the files in the app were using tabs rather than spaces for indents (must of changed the sublime settings inadvertently). Those changes were incorporated into this PR.

However, the size is primarily due to (1) this is a new app and in many instances, the work going forward relies the work that immediately preceded it. In this case, the base CsvFile and the first derived instance WeamsCsvFile - I need the CsvFile resource for the WeamsCsvFile. (2) a need to develop a better solution that was more workable, necessitating a quick but major rewrite, and there was no way I could figure to do that piecemeal in a fashion that would make a several incomplete code reviews for smaller changes any more palatable that one large change (kind of like ripping off bandaid slowly vs all at once); and most importantly (3) the time constraints - right now I am the only developer and must rely on you (and last night Albert) to do a code review. This means that I would have to wait several hours for the CsvFile stuff to get approved before I could work on the derived WeamsCsvFile. 

I made a decision, and it was probably the wrong one. I am not trying to be difficult, but am looking to wrap up this project by the end of next week or so that the Charter slip isn't too bad.

Going forward, I am scheduled to churn out 2 additional CsvFile resources a day - controllers, models, views, and specs and 2 intermediate data resources (tables that hold the processed data from a CSV) a day over the next several days. We need to come to some agreement on how the review goes for this. Just let me know how you want it done, and I will do it. I had promised Stacy the Crosswalk resources (VaCrosswalk, VaCrosswalkCsvFile, and the models, controlers, specs, and views) for today. But it is all based on acceptance of the CsvFile stuff in this PR. 

Sorry if I seem whiny, and I know you are probably far busier than I, but I really don't know how to keep to this schedule and produce PRs in an acceptable fashion for quick turnaround.

File: app/controllers/csv_files_controller.rb (hunk 1 above)
Comment (plusjeff) [inline app/controllers/csv_files_controller.rb]: Whoops! Sorry for the ambiguity. I meant `STI`. :smile_cat:

File: app/models/csv_files/csv_file.rb (hunk 5 above)
Comment (plusjeff) [inline app/models/csv_files/csv_file.rb]: Sounds good! Thanks for the explanation. Could you leave that in comment next to the one that is sticking around?

File: app/views/csv_files/_csv_file.html.erb (hunk 6 above)
Comment (plusjeff) [inline app/views/csv_files/_csv_file.html.erb]: Gotcha. @rickleegit: Was does your design wisdom say we should name this?

**Comment by jeffmaher** [comment]
Thanks for the context @mphprogrammer ! Working on finishing reviewing today. Some suggestions on keeping the PR size down:
- When formatting changes occur (like tabs/spaces), submit those in their own PR
- Any time you have a small piece of testable functionality, submit a PR. This shouldn't stop from building a future branch (you can always branch off your PR branch while you wait for a code reviewer)

Definitely hear you on the slowness of getting help with the reviews - that definitely sucks. With the team changes that are on the slate for next sprint, you'll have other developers to help out on the way.

File: config/initializers/preload_sti_models.rb [hunk 8]
```diff
@@ -1,10 +1,10 @@
-###############################################################################
-## Preloads all sti classes in development, rather than lazy loading. The 
-## parent class needs to recognize its children ASAP. (mph)
-###############################################################################
+##############################################################################
+# Preloads all sti classes in development, rather than lazy loading. The
```
Comment (plusjeff) [inline config/initializers/preload_sti_models.rb]: Comment nitpick: Put in "Single Table Inheritance".

File: db/migrate/20160311045126_create_csv_files.rb [hunk 9]
```diff
@@ -1,11 +1,9 @@
-class CreateRawFiles < ActiveRecord::Migration
+class CreateCsvFiles < ActiveRecord::Migration
```
Comment (plusjeff) [inline db/migrate/20160311045126_create_csv_files.rb]: Did GitHub goof on this? Normally migration files don't move.

File: spec/controllers/csv_files_controller_spec.rb [hunk 10]
```diff
@@ -0,0 +1,265 @@
+require 'rails_helper'
```
Comment (plusjeff) [inline spec/controllers/csv_files_controller_spec.rb]: Yay! Tests! :smile_cat:

File: spec/controllers/csv_files_controller_spec.rb [hunk 11]
```diff
@@ -0,0 +1,265 @@
...
+
+    before(:each) do
+      create :weams_csv_file
+      @csv = WeamsCsvFile.first
+
+      get :index
+    end
+
+    it "populates an array of csvs" do
+      expect(assigns(:csv_files)).to include(@csv)
```
Comment (plusjeff) [inline spec/controllers/csv_files_controller_spec.rb]: This is just me not being super savvy on the syntax. For my own curiosity, what does this line do and what magic happens that verifies that the array of CSVs is populated.

File: spec/helpers/weams_helper_spec.rb [hunk 12]
```diff
@@ -0,0 +1,15 @@
+require 'rails_helper'
```
Comment (plusjeff) [inline spec/helpers/weams_helper_spec.rb]: Is this file needed? Does it do anything?

**Comment by jeffmaher** [comment]
Did a skimmed review of the rest of the code. Had a few questions, so assigning over to you @mphprogrammer . Pop back to me when ready for re-review and/or merging.

File: app/controllers/csv_files_controller.rb (hunk 1 above)
Comment (mphprogrammer) [inline app/controllers/csv_files_controller.rb]: Removing, no longer need the STI hash.

File: db/migrate/20160311045126_create_csv_files.rb (hunk 9 above)
Comment (mphprogrammer) [inline db/migrate/20160311045126_create_csv_files.rb]: That was a tab to spaces change.

File: app/views/csv_files/_csv_file.html.erb (hunk 6 above)
Comment (mphprogrammer) [inline app/views/csv_files/_csv_file.html.erb]: I will call it "Remove" for now, its a simple change that we can redo at a later date.

File: config/initializers/preload_sti_models.rb (hunk 8 above)
Comment (mphprogrammer) [inline config/initializers/preload_sti_models.rb]: Done

File: app/models/csv_files/csv_file.rb (hunk 5 above)
Comment (mphprogrammer) [inline app/models/csv_files/csv_file.rb]: Done

File: spec/controllers/csv_files_controller_spec.rb (hunk 10 above)
Comment (mphprogrammer) [inline spec/controllers/csv_files_controller_spec.rb]: Woo Hoo!

File: spec/controllers/csv_files_controller_spec.rb (hunk 11 above)
Comment (mphprogrammer) [inline spec/controllers/csv_files_controller_spec.rb]: This line checks that the controller populates the variable :csv_files. The test preloads a single csv_file into the database, and the spec ensures that value is loaded into the array.

File: spec/helpers/weams_helper_spec.rb (hunk 12 above)
Comment (mphprogrammer) [inline spec/helpers/weams_helper_spec.rb]: The file has been removed.

**Comment by mphprogrammer** [comment]
I've gone through the changes, passing back to you.
